File: c/problem_48.c
#include "problem_48.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

typedef struct Node48 { char val; struct Node48 *left, *right; } Node48;

static Node48 nodes48[128];
static int node_idx48 = 0;

static Node48 *new_node48(char v) {
    Node48 *n = &nodes48[node_idx48++];
    n->val = v; n->left = n->right = NULL;
    return n;
}

static Node48 *build(const char *pre, int prelen, const char *ino, int inolen) {
    if (prelen == 0) return NULL;
    Node48 *root = new_node48(pre[0]);
    int idx = 0;
    while (idx < inolen && ino[idx] != pre[0]) idx++;
    root->left  = build(pre+1,       idx,        ino,      idx);
    root->right = build(pre+1+idx,   prelen-1-idx, ino+idx+1, inolen-idx-1);
    return root;
}

static void inorder48(Node48 *n, char *out, int *i) {
    if (!n) return;
    inorder48(n->left, out, i);
    out[(*i)++] = n->val;
    inorder48(n->right, out, i);
}

void problem48(void) {
    printf("Problem 48:\n");
    node_idx48 = 0;
    const char *pre = "abdecfg";
    const char *ino = "dbeafcg";
    Node48 *root = build(pre, 7, ino, 7);
    char out[8] = {0}; int i = 0;
    inorder48(root, out, &i);
    assert(strcmp(out, ino) == 0);
    printf("All Tests Passed\n");
}

File: typescript/problem_48.ts
class TreeNode48ts {
    constructor(public val: string, public left: TreeNode48ts | null = null, public right: TreeNode48ts | null = null) {}
}

export function buildTree(preorder: string[], inorder: string[]): TreeNode48ts | null {
    if (!preorder.length) return null;
    const root = new TreeNode48ts(preorder[0]);
    const idx = inorder.indexOf(preorder[0]);
    root.left  = buildTree(preorder.slice(1, 1 + idx), inorder.slice(0, idx));
    root.right = buildTree(preorder.slice(1 + idx), inorder.slice(idx + 1));
    return root;
}

function inorderTs(root: TreeNode48ts | null): string[] {
    if (!root) return [];
    return [...inorderTs(root.left), root.val, ...inorderTs(root.right)];
}

export function problem48(): void {
    console.log("Problem 48:");
    const pre = ["a","b","d","e","c","f","g"];
    const ino = ["d","b","e","a","f","c","g"];
    if (JSON.stringify(inorderTs(buildTree(pre, ino))) !== JSON.stringify(ino))
        throw new Error("assertion failed");
    console.log("All Tests Passed");
}
